Tests fail because C++ compiler has extra flag enabled in MacOS

Dafny commit 0e4b2bbd7f4adcb6fcc23a11e28bdbbcabbf5cee
macOS 12.3.1 (on Apple Silicon, in case that matters)
```
g++ --version
Apple clang version 13.1.6 (clang-1316.0.21.2.3)
Target: arm64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

This looks like a close repeat of old issue #1306.

Some integration tests fail because the C++ compiler emits warnings that throw off the test result diff. The warning is emitted by `-Wunused-but-set-variable`, which in [Compilers/Compiler-cpp.cs:2429](https://github.com/dafny-lang/dafny/blob/master/Source/Dafny/Compilers/Compiler-cpp.cs#L2429) is disabled for Windows and Linux but remains enabled for MacOS. 


Failed test output:
```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:01:03.20]     c++/const.dfy [FAIL]
  Failed c++/const.dfy [4 s]
  Error Message:
   System.Exception : Command returned non-zero exit code (1): DiffCommand TestFiles/LitTests/LitTest/c++/const.dfy.expect /Users/mija/repos/dafny/dafny/Source/IntegrationTests/bin/Debug/net6.0/TestFiles/LitTests/LitTest/c++/Output/const.dfy.tmp
Output:
Xunit.Sdk.AssertActualExpectedException: AssertEqualWithDiff() Failure
Diff (changing expected into actual):

 Dafny program verifier finished with 1 verified, 0 errors
+TestFiles/LitTests/LitTest/c++/const.cpp:16:10: warning: variable '_0_b' set but not used [-Wunused-but-set-variable]
+    bool _0_b;
+         ^
+1 warning generated.


Expected:
Dafny program verifier finished with 1 verified, 0 errors

Actual:
Dafny program verifier finished with 1 verified, 0 errors
TestFiles/LitTests/LitTest/c++/const.cpp:16:10: warning: variable '_0_b' set but not used [-Wunused-but-set-variable]
    bool _0_b;
         ^
1 warning generated.

   at XUnitExtensions.AssertWithDiff.Equal(String expected, String actual) in /Users/mija/repos/dafny/dafny/Source/XUnitExtensions/AssertWithDiff.cs:line 28
   at XUnitExtensions.Lit.DiffCommand.Execute(ITestOutputHelper outputHelper, TextReader inputReader, TextWriter outputWriter, TextWriter errorWriter) in /Users/mija/repos/dafny/dafny/Source/XUnitExtensions/Lit/DiffCommand.cs:line 35
Error:

  Stack Trace:
     at XUnitExtensions.Lit.LitTestCase.Execute(ITestOutputHelper outputHelper) in /Users/mija/repos/dafny/dafny/Source/XUnitExtensions/Lit/LitTestCase.cs:line 94
   at XUnitExtensions.Lit.LitTestCase.Run(String filePath, LitTestConfiguration config, ITestOutputHelper outputHelper) in /Users/mija/repos/dafny/dafny/Source/XUnitExtensions/Lit/LitTestCase.cs:line 55
   at IntegrationTests.LitTests.LitTest(String path) in /Users/mija/repos/dafny/dafny/Source/IntegrationTests/LitTests.cs:line 157
  Standard Output Messages:
 Executing command: dotnet /Users/mija/repos/dafny/dafny/Source/IntegrationTests/bin/Debug/net6.0/Dafny.dll /countVerificationErrors:0 /vcsCores:2 /useBaseNameForFileName /compileVerbose:0 /errorTrace:0 /timeLimit:300 /compile:3 /spillTargetCode:2 /compileTarget:cpp TestFiles/LitTests/LitTest/c++/const.dfy > /Users/mija/repos/dafny/dafny/Source/IntegrationTests/bin/Debug/net6.0/TestFiles/LitTests/LitTest/c++/Output/const.dfy.tmp
 Executing command: DiffCommand TestFiles/LitTests/LitTest/c++/const.dfy.expect /Users/mija/repos/dafny/dafny/Source/IntegrationTests/bin/Debug/net6.0/TestFiles/LitTests/LitTest/c++/Output/const.dfy.tmp
 Previous command results:
 Output:

 Dafny program verifier finished with 1 verified, 0 errors
 TestFiles/LitTests/LitTest/c++/const.cpp:16:10: warning: variable '_0_b' set but not used [-Wunused-but-set-variable]
     bool _0_b;
          ^
 1 warning generated.

 Error:


```

--- a/Source/Dafny/Compilers/Compiler-cpp.cs
+++ b/Source/Dafny/Compilers/Compiler-cpp.cs
@@ -2424,10 +2424,22 @@ namespace Microsoft.Dafny.Compilers {
       Contract.Assert(assemblyLocation != null);
       var codebase = System.IO.Path.GetDirectoryName(assemblyLocation);
       Contract.Assert(codebase != null);
-      var warnings = "-Wall -Wextra -Wpedantic -Wno-unused-variable -Wno-deprecated-copy -Wno-unused-label -Wno-unused-but-set-variable -Wno-unknown-warning-option";
-      var args = warnings + $" -g -std=c++17 -I {codebase} -o {ComputeExeName(targetFilename)} {targetFilename}";
+      var args = new List<string> {
+                       "-Wall",
+                       "-Wextra",
+                       "-Wpedantic",
+                       "-Wno-unused-variable",
+                       "-Wno-deprecated-copy",
+                       "-Wno-unused-label",
+                       "-Wno-unused-but-set-variable",
+                       "-Wno-unknown-warning-option",
+                        "-g",
+                        "-std=c++17",
+                        "-I", $"{codebase}",
+                        "-o", $"{ComputeExeName(targetFilename)}",
+                        $"{targetFilename}" };
       compilationResult = null;
-      var psi = new ProcessStartInfo("g++", args) {
+      var psi = new ProcessStartInfo("g++", string.Join(" ", args)) {
         CreateNoWindow = true,
         UseShellExecute = false,
         RedirectStandardOutput = true,

--- a/Source/Dafny/Compilers/Compiler-cpp.cs
+++ b/Source/Dafny/Compilers/Compiler-cpp.cs
@@ -2435,9 +2435,9 @@ namespace Microsoft.Dafny.Compilers {
                        "-Wno-unknown-warning-option",
                         "-g",
                         "-std=c++17",
-                        "-I", $"{codebase}",
+                        "-I", codebase,
                         "-o", $"{ComputeExeName(targetFilename)}",
-                        $"{targetFilename}" };
+                        targetFilename };
       compilationResult = null;
       var psi = new ProcessStartInfo("g++", string.Join(" ", args)) {
         CreateNoWindow = true,

--- a/Source/Dafny/Compilers/Compiler-cpp.cs
+++ b/Source/Dafny/Compilers/Compiler-cpp.cs
@@ -2436,7 +2436,7 @@ namespace Microsoft.Dafny.Compilers {
                         "-g",
                         "-std=c++17",
                         "-I", codebase,
-                        "-o", $"{ComputeExeName(targetFilename)}",
+                        "-o", ComputeExeName(targetFilename),
                         targetFilename };
       compilationResult = null;
       var psi = new ProcessStartInfo("g++", string.Join(" ", args)) {

--- a/Source/Dafny/Compilers/Compiler-cpp.cs
+++ b/Source/Dafny/Compilers/Compiler-cpp.cs
@@ -2424,26 +2424,27 @@ namespace Microsoft.Dafny.Compilers {
       Contract.Assert(assemblyLocation != null);
       var codebase = System.IO.Path.GetDirectoryName(assemblyLocation);
       Contract.Assert(codebase != null);
-      var args = new List<string> {
-                       "-Wall",
-                       "-Wextra",
-                       "-Wpedantic",
-                       "-Wno-unused-variable",
-                       "-Wno-deprecated-copy",
-                       "-Wno-unused-label",
-                       "-Wno-unused-but-set-variable",
-                       "-Wno-unknown-warning-option",
-                        "-g",
-                        "-std=c++17",
-                        "-I", codebase,
-                        "-o", ComputeExeName(targetFilename),
-                        targetFilename };
       compilationResult = null;
-      var psi = new ProcessStartInfo("g++", string.Join(" ", args)) {
+      var psi = new ProcessStartInfo("g++") {
         CreateNoWindow = true,
         UseShellExecute = false,
         RedirectStandardOutput = true,
-        RedirectStandardError = true
+        RedirectStandardError = true,
+        ArgumentList = {
+          "-Wall",
+          "-Wextra",
+          "-Wpedantic",
+          "-Wno-unused-variable",
+          "-Wno-deprecated-copy",
+          "-Wno-unused-label",
+          "-Wno-unused-but-set-variable",
+          "-Wno-unknown-warning-option",
+          "-g",
+          "-std=c++17",
+          "-I", codebase,
+          "-o", ComputeExeName(targetFilename),
+          targetFilename
+        }
       };
       var proc = Process.Start(psi);
       while (!proc.StandardOutput.EndOfStream) {

--- a/Source/Dafny/Compilers/Compiler-cpp.cs
+++ b/Source/Dafny/Compilers/Compiler-cpp.cs
@@ -2424,12 +2424,7 @@ namespace Microsoft.Dafny.Compilers {
       Contract.Assert(assemblyLocation != null);
       var codebase = System.IO.Path.GetDirectoryName(assemblyLocation);
       Contract.Assert(codebase != null);
-      var warnings = "-Wall -Wextra -Wpedantic -Wno-unused-variable -Wno-deprecated-copy -Wno-unused-label";
-      if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) {
-        warnings += " -Wno-unused-but-set-variable";
-      } else {
-        warnings += " -Wno-unknown-warning-option";
-      }
+      var warnings = "-Wall -Wextra -Wpedantic -Wno-unused-variable -Wno-deprecated-copy -Wno-unused-label -Wno-unused-but-set-variable -Wno-unknown-warning-option";
       var args = warnings + $" -g -std=c++17 -I {codebase} -o {ComputeExeName(targetFilename)} {targetFilename}";
       compilationResult = null;
       var psi = new ProcessStartInfo("g++", args) {

